tenseal/cpp/tensors/ckkstensor.cpp

```diff
@@ -129,22 +129,165 @@ shared_ptr<CKKSTensor> CKKSTensor::power_inplace(unsigned int power) {
     return shared_from_this();
 }
 
+void CKKSTensor::perform_op(seal::Ciphertext& ct, seal::Ciphertext other,
+                            OP op) {
+    this->auto_same_mod(other, ct);
+    switch (op) {
+        case OP::ADD:
+            this->tenseal_context()->evaluator->add_inplace(ct, other);
+            break;
+        case OP::SUB:
+            this->tenseal_context()->evaluator->sub_inplace(ct, other);
+        case OP::MUL:
+            this->tenseal_context()->evaluator->multiply_inplace(ct, other);
+            this->auto_relin(ct);
+            this->auto_rescale(ct);
+            break;
+        default:
+            throw invalid_argument("operation not defined");
+    }
+}
+
+void CKKSTensor::perform_plain_op(seal::Ciphertext& ct, seal::Plaintext other,
+                                  OP op) {
+    this->auto_same_mod(other, ct);
+    switch (op) {
+        case OP::ADD:
+            this->tenseal_context()->evaluator->add_plain_inplace(ct, other);
+            break;
+        case OP::SUB:
+            this->tenseal_context()->evaluator->sub_plain_inplace(ct, other);
+        case OP::MUL:
+            this->tenseal_context()->evaluator->multiply_plain_inplace(ct,
+                                                                       other);
+            this->auto_relin(ct);
+            this->auto_rescale(ct);
+            break;
+        default:
+            throw invalid_argument("operation not defined");
+    }
+}
+
+shared_ptr<CKKSTensor> CKKSTensor::op_inplace(
+    const shared_ptr<CKKSTensor>& operand, OP op) {
+    // TODO implement broadcasting
+    if (this->_shape != operand->_shape) {
+        // TODO provide a better message (what are the shapes)
+        throw invalid_argument("Operands don't have the same shape");
+    }
+
+    size_t n_jobs = this->tenseal_context()->dispatcher_size();
+
+    auto worker_func = [&](size_t start, size_t end) -> bool {
+        for (size_t i = start; i < end; i++) {
+            this->perform_op(this->_data[i], operand->_data[i], op);
```

Comment:
As far as I see, perform_op can throw exceptions.

We shouldn't not throw exceptions in the workers, as it might lead to undefined behavior in the other threads. I think we had a similar crash here.  https://github.com/OpenMined/TenSEAL/pull/151